src/ipyradiant/visualization/vis_tool.py

```diff
@@ -0,0 +1,114 @@
+    cyto_widget = T.Instance(CytoscapeWidget, allow_none=True)
+    nodes = T.List()
+    click_output = T.Instance(W.Output)
+    click_output_box = T.Instance(W.VBox)
+    selected_node = T.Unicode(allow_none=True)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.output_title = W.HTML("<h3>Output from Clicks</h3>")
+        self.layout = W.Layout(grid_template_columns="60% 40%", height="500px")
+        self.click_output_box.children = [self.output_title, self.click_output]
+        self.children = [self.cyto_widget, self.click_output_box]
+
+    @T.default("click_output")
+    def _make_click_output(self):
+        return W.Output()
+
+    @T.default("nodes")
+    def _make_default_nodes(self):
+        return []
+
+    @T.default("click_output_box")
+    def _make_default_click_output_box(self):
+        return W.VBox()
+
+    @T.default("node_selector_box")
+    def _make_node_selector_box(self):
+        return W.VBox()
+
+    @T.default("cyto_widget")
+    def _make_default_cyto_widget(self):
+        cyto_widget = CytoscapeWidget()
+        cyto_widget.on("node", "click", self.log_node_clicks)
+        cyto_widget.on("edge", "click", self.log_edge_clicks)
+        cyto_widget.set_style([default_node, default_edge])
+        return cyto_widget
+
+    def log_node_clicks(self, node):
+        with self.click_output:
+            print(f"node clicked: {node['data']}")
+            print("-------------------------------")
+
+    def log_edge_clicks(self, edge):
+        with self.click_output:
+            print(f"edge clicked:")
+            print(f'edge source: {edge["data"]["source"]}')
+            print(f'edge target: {edge["data"]["target"]}')
+            print("-------------------------------")
+
+    @T.observe("graph")
+    def update_cyto_widget_graph(self, change):
+        # TODO configure vis tool with replace=True/False
+        # remove old nodes (which removed other links, e.g. edges)
+        for node in list(self.cyto_widget.graph.nodes):
+            self.cyto_widget.graph.remove_node(node)
+        assert len(self.cyto_widget.graph.nodes) == 0, "OHNO"
+        new_json = build_cytoscape_json(change.new)
+        self.cyto_widget.graph.add_graph_from_json(new_json, directed=True)
+
+
+# # TODO:
+# Why are there missing edges, etc in the graph and why does it look wrong
+# do we want the check for object not being literal ???
+
+
+def build_cytoscape_json(graph: Graph):
+    # collect uris & edges
+    element_set = set()
+    edges = []
+    for i, (s, p, o) in enumerate(graph):
+        if p != RDF.type:
+            if isinstance(s, URIRef):
+                element_set.add(s)
+            if isinstance(o, URIRef):
+                element_set.add(o)
+            if not isinstance(o, Literal):
+                edges.append(
+                    {"source": s, "target": o, "label": f"{Path(p).name}",}
+                )
+        # edges.append(
```

Comment:
Is this commented code still needed?

Reply:
same as  other reply with the ##TODO... do we want to add all edges or only ones with objects that aren't literals?